First version

EECOLOR opened the pull request:
This pull request should lead to the first version. Any comments are welcome.

- [x] Make it work (it has some problems caused by refactoring)
- [x] ~~Split 'basic' and 'route map' examples~~
- [x] Add nested example (utilizing `basePath`)
- [x] Provide access to the parent routes in 'route map' version
- [x] Add smarter `*` mapping in 'route map'  (allows for more granual `not found` checking)
- [x] Add `useRelativePickFromRouteMap` (or something equivalent)
- [x] Add documentation
- [x] ~~Deal with `http` and `target='_blank' (`rel="noopener noreferrer"`)~~
- [x] Write tests (only for the easily testable functionality at first)
- [x] Disable default [`jsx-key`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-key.md) and replace with a version that allows 'key-less' react elements in array literals (99% of the time these are tuples)
- [ ] More realistic examples

EECOLOR committed:
initial sketch

EECOLOR committed:
remove rogue files

EECOLOR committed:
code cleanup

EECOLOR committed:
improve Link

EECOLOR committed:
improve code hinting during development

EECOLOR committed:
change `pick` to `routes`  and improve typing

EECOLOR committed:
add `route` to `useRouting`

EECOLOR committed:
first sketch

please do not look the code

EECOLOR committed:
more sketching

Explored exhaustive /* routes based on children. It can be done. Most
likely candidate is making `extractPath` return an array of all possible
child routes (separated with /*). That requires a change in matching
where /*abcdefg is treated differently from /*. We probably also need
to change scoring to include length for equal scores (or use length as
a base of the score).

EECOLOR committed:
add relative pick

EECOLOR committed:
first sketch with reverse routing

EECOLOR committed:
more sketching, making the language switch work

EECOLOR committed:
Simplification of the library

- Route map is now a default (optional) integration
- We explicitly require LocationProvider to prevent 'magic'
- No more implicit LocationProviders

EECOLOR committed:
Update TODO's

EECOLOR committed:
Merge pull request #2 from kaliberjs/route-map

Route map

peeke commented:
First off, looks great, nice and clean 👍 

I've been hacking about, trying to get page transitions to work. The code below is something along the lines of what I've tried. I've summarized the main takeaways below.

```js
export default function App({ initialLocation }) {
  const { pick } = useRouting({ initialLocation })

  return (
    <>
      <Navigation />
      {pick(['*', <AnimatedRoutes />])} // Wrapping in a Provider would be nicer here
    </>
  )
}

function AnimatedRoutes() {
  const { pick } = useRouting()
  const location = useLocation()
  const transitions = useTransition(location, location => location.key, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0, position: 'absolute' },
  })

  return transitions.map(({ item: location, key, props }) => (
    <animated.div style={props} {...{ key }}>
      {pick(
        [
          ['', <Home />],
          ['articles', <Articles />],
          ['articles/:articleId/*', params => <Article {...{ params }} />],
          ['*', params => <NotFound {...{ params }} />],
        ],
        location // Or maybe wrap this whole element in a <LocationProvider />?
      )}
    </animated.div>
  ))
}
```

- For animation purposes, it would like to be able to provide my own location to `pick`. Or provide it through a LocationProvider. This way,  I can provide a transitioning component with an old location, while the transition is in progress. Also check the last example here: https://www.react-spring.io/docs/hooks/use-transition
- If my first use of the location is not routing through pick, but using it through `useLocation`, it would be nice if I could manually provide my application with the location. E.g., be exporting the `LocationProvider`.

EECOLOR commented:
Great feedback. The example you posted however should not work as the library now requires a `LocationProvider` (see examples).

> E.g., be exporting the `LocationProvider`.

In this version it is exported.

I'll look at the rest of your comments as soon as I have time.

peeke commented:
> Great feedback. The example you posted however should not work as the library now requires a `LocationProvider` (see examples).

It doesn't. I've added it to show what I've tried, might provide some insight.

EECOLOR committed:
Refactor and add tests

Trying a different route ;-) where `routeMap` is the only option.

EECOLOR committed:
Move towards routeMap only routing

It seems better to only have 1 way of doing things. This simplifies code
and provides extra featues (fallback to not found, even after dynamic
routes).

It also provides some other benefits because we can have absolute paths
in a relative fashion.

EECOLOR committed:
Prefil available params in reverse routes

EECOLOR committed:
Internal refactoring

EECOLOR committed:
Fix advanced example

Incidentally added language support.
Some features still missing.
Needs a bit of polish, but ran out of time

EECOLOR committed:
Correctly handle localization errors

Some refactoring as well

EECOLOR committed:
Working towards useful types

EECOLOR committed:
Refactor inner workings

- Group context by rate of change
- Only a single match for each location change
- More sensible hooks

EECOLOR committed:
Add support for route data

Also in this pull request:
- More advanced type definitions for routeMap
- Refactoring of route definition

EECOLOR committed:
Typescript related tweaks

EECOLOR commented:
https://github.com/kaliberjs/routing/pull/4#discussion_r589330108

EECOLOR committed:
Small simplification in example

EECOLOR committed:
Merge pull request #4 from kaliberjs/different-direction

Different direction

EECOLOR commented:
@peeke Much has changed, can you check what is missing for animations? We might want to create a separate branch and pull request to discuss things there.

EECOLOR committed:
Add readme

larsvankleef commented on example/src/advanced/routeMap.js:
```diff
@@ -0,0 +1,59 @@
+  return { price: 10 }
+}
+
+async function second(x) { return new Promise(resolve => setTimeout(resolve, x * 1000)) }
```
Dit is niet de juiste functie naam voor wat hij doet. Zou hier kiezen voor `wait` of `delay`

EECOLOR replied:
Het leest wel lekker waar het gebruikt wordt: `await second(1)`

EECOLOR replied:
```
await resolveAfter({ seconds: 1 })
```

larsvankleef commented on example/src/App.js:
```diff
@@ -0,0 +1,93 @@
+}
+
+function Page() {
+  const { routes } = useRouting()
```
Zou dit niet `{ router }` kunnen heten? Dat voelt wel logischer met wat de functie doet en dan is de `routeMap` misschien ook wat logischer. Je gebruikt dan de `routeMap` variabele in de `router`.

```
router(
  [home, <Home />],
  [articles, <Articles />]
  [articles.article, params => <Article {...{ params }} />],
  [notFound, params => <NotFound {...{ params }} />],
)
```

EECOLOR replied:
Ik snap wat je zegt. Naar aanleiding van ons gesprek heb ik gekeken hoe het heet bij React Router 6 en daar heet het ook `routes`. De `routeMap` wordt hier trouwens niet gebruikt, maar wel routes.

In React Router 6 zou het er zo uitzien:

```jsx
<Routes>
  <Route path="/" element={<Home />} />
  <Route path="/articles" element={<Home />} />
  ...
```

Dus in mijn hoofd past dat perfect.

EECOLOR replied:
Het is veranderd naar `matchRoute` en `matchRoutes`

larsvankleef commented on example/src/App.js:
```diff
@@ -0,0 +1,93 @@
+        <Link to={tab2()}>Tab2</Link>
+      </div>
+      <div>
+        {routes(
```
Misschien is het wat verbose, maar ik zou in dit voorbeeld even de strings die nu gebruikt worden in `routes` vervangen door componenten. Dit omdat het nu heel gek is dat er ineens strings in staan terwijl je hiervoor geleerd hebt dat je componenten er in moet zetten.

EECOLOR replied:
> terwijl je hiervoor geleerd hebt dat je componenten er in moet zetten

Het punt is dat het niet uit maakt wat je er in zet. Het voorbeeld van strings geeft juist aan dat het prima functioneert zonder componenten. Dus dat je het ook op pekken kunt gebruiken waar alleen een titel veranderd op basis van de route.

Misschien bedoel je iets anders, dan moeten we het even bespreken.

larsvankleef commented on example/src/App.js:
```diff
@@ -0,0 +1,93 @@
+          [tab2, 'Tab 2'],
+        ) || 'not matched'}
+      </div>
+      {route(tab1, <div>Side bar for tab 1</div>)}
```
Ik snap niet waar voor dit er staat?

EECOLOR replied:
We tonen een side bar wanneer `tab1` geselecteerd is. Ik denk dat het een goed idee is om hier een `<aside>` te gebruiken. Het voorbeeld is belangrijk om te laten zien dat een component op verschillende manieren op een route kan reageren. Dit component rendert twee trees die onafhankelijk van elkaar op routes reageren.

larsvankleef commented on example/src/advanced/App.js:
```diff
@@ -0,0 +1,158 @@
+function Page() {
+  const { route, routes } = useRouting()
+  const { language } = useRoutes()
+  const { home, articles, notFound } = language
```
Als je dit leest zonder dat je weet hoe de `routeMap` er uit ziet deze code er wel gek uit. Omdat `language` niet voelt als iets waar routes in zitten.

EECOLOR replied:
Ik snap wat je zegt. Zou dit beter zijn?

```js
const { language, language: { home, articles, notFound } } = useRoutes()
```

Ik denk wel dat we een naamgevingsprobleem hebben (zoals je eerder al aangaf):

```js
const { routes } = useRouting()
const routes = useRoutes()
...
routes(
  [routes.language, ...]
)
```
Dat moet dan dit worden:

```js
const routing = useRouting()
const routes = useRoutes()
...
routing.routes(
  [routes.language, ...]
)
```

Geeft me trouwens wel een idee, haha. `useRouting` en `useRoutes` worden vaak samen gebruikt, we zouden dus `route` en `routes` een dubbele functie kunnen geven, zoiets:

```js
const { routes } = useRouting()

routes(
  routes.home,
  routes.articles,
  ...
)
```

EECOLOR replied:
Nee Erik, dat maakt het moeilijker, het is nu `matchRoute` en `matchRoutes` geworden zodat we `routes` vrij kunnen gebruiken. In de voorbeelden is het nu ook `routes.myRoute`

larsvankleef commented on example/src/advanced/App.js:
```diff
@@ -0,0 +1,158 @@
+function Articles() {
+  const { article, list } = useRoutes()
+  const language = useLanguage()
+  const { title } = useRoute().data
```
Zouden we hier niet een extra functie voor kunnen maken? `useRouteData`  omdat we dat nu ook hebben voor `useAsyncRouteData` Voelt wel fijn om dan `useRouteData` te gebruiken omdat het dan consistenter is. Of misschien een functie? Want zou niet eigenlijk alle data uit de `useRouteData` terug kunnen komen ongeacht of het async is of niet.

Ik snap ook niet helemaal waarom ik de `route` moet mee geven bij de `useAsyncRouteData` Kan die dat niet zelf uit vogelen?

EECOLOR replied:
`useRouteData` is een goed idee.

Je hoeft de `route` niet mee te geven bij `useAsyncRouteData`, dat is alleen wanneer je een andere dan de huidige route wilt meegeven.

EECOLOR replied:
> Want zou niet eigenlijk alle data uit de useRouteData terug kunnen komen ongeacht of het async is of niet.

Dat kan niet, kunnen we eventueel nog even bespreken.

larsvankleef commented on example/src/advanced/App.js:
```diff
@@ -0,0 +1,158 @@
+}
+
+function Sidebar({ article }) {
+  const { price } = useAsyncRouteData({ price: 0 }, { extraArgs: { article } })
```
Zouden we `extraArgs` niet gewoon open kunnen laten? Waardoor je eerste default return waardes hebt en daarna context die je wil mee geven aan de functie in `routeMap`

Zou ook al kunnen werken als `extraArgs` wordt hernoemend naar `context`. Dit is de context waarin je de functie wil uit voeren.

`useAsyncRouteData('default return values', 'context')`

EECOLOR replied:
Ik ben het met je eens dat `extraArgs` geen fijne naam is. En (naar aanleiding van ons gesprek) ook dat de 'example' code vaak gekopieerd gaat worden en dus pico bello zou moeten zijn.

`context` vind ik geen goede naam. Wat je daar meegeeft zijn argumenten voor de `data` functie wat geen `params` zijn van je `path`.

De reden dat het in een object zit is omdat beide argumenten optioneel zijn.

EECOLOR replied:
Na er nog eens over nagedacht te hebben denk ik dat ik dit gehele stuk moet refactoren. Ik dacht: het is een example, dus het boeit allemaal niet zo veel.

Maar ik zou moeten denken: 'dit is een example, dit gaan mensen kopieren'. Ik zal hier dus een `react-query` variant van moeten maken.

EECOLOR replied:
Ik ga hier nu geen tijd meer in stoppen totdat ik ervaring heb met een realistische situatie.

larsvankleef commented on example/src/advanced/App.js:
```diff
@@ -0,0 +1,158 @@
+        </div>
+      )}
+      <div>
+        {routes(
```
Hier is het misschien nog wel veel duidelijker dat het `router` gaat heten. Dan is het ook logische dat je daar onder `route()` hebt staan voor het renderen van een specifieke `route`

EECOLOR replied:
Ik denk dat we nog even goed moeten kijken naar de naamgeving. Hier is het misschien slim om de andere developers ook bij te betrekken.

larsvankleef commented on example/src/advanced/machinery/routeData.js:
```diff
@@ -0,0 +1,79 @@
+  const dataForRef = React.useRef(state === defaultValue ? null : routeId)
+  const { params } = useMatch()
+
+  const getData = route.data
+  const props = { ...params, ...extraArgs }
+
+  React.useEffect(
+    () => {
+      if (dataForRef.current === routeId) return
+      dataForRef.current = routeId
+
+      fetchData().catch(e => console.error(e))
+      return () => { dataForRef.current = null }
+
+      async function fetchData() {
+        const result = await getData(props)
+        if (dataForRef.current !== routeId) return
+        setState(result)
+      }
+    },
+    [routeId, getData, props]
+  )
+
+  return state
+}
+
+function useSingleUseProperties(initialData) {
+  return React.useMemo(
+    () => Object.entries(initialData).reduce(
+      (result, [k, v]) => (Object.defineProperty(result, k, { get: getOnce(v) })),
+      {}
+    ),
+    [initialData]
+  )
+}
+
+function getOnce(data) {
+  let used = false
+  return () => used ? undefined : (used = true, data) // eslint-disable-line no-return-assign
+}
```
Vraag mij af in hoevere deze functie `machinery` functies zijn die verschillen per project. Het ophalen van data is eigenlijk altijd hetzelfde? in de zin van wat deze functies doen. Ze voeren de functues uit die in je `routeMap` staan die per project kunnen verschillen.

EECOLOR replied:
`machinery` betekent: bevat geen domein specifieke code. Niets in deze code verwijst naar het domein van dit project. Het stelt wel eisen aan de omgeving: `data` moet een `function` zijn die een `object` accepteert. Daarnaast wordt er vanuit gegaan dat er een `@kaliber/routing` context beschikbaar is.

De reden dat dit in machinery zit en niet in de library is tweeledig:

1. Het past niet echt in de library, is een ander onderwerp
2. We weten nog niet goed wat handig is

Echter, in lijn met je vorige opmerking over copy-paste van examples mag dit misschien nog iets meer aandacht krijgen. Misschien slim als we ook een voorbeeldje met Sanity maken zodat we wat meer gevoel krijgen bij wat handig is.

larsvankleef commented on example/src/advanced/index.html.js:
```diff
@@ -0,0 +1,50 @@
+    )
+
+    return 'route' in result
+      ? { ...result, data: { routeData: await fetchRouteData(result.route, result.params) } }
```
Vraagje, Hoe werkt het data fetch als je data specifiek is aan of een gebruiker. Hoe zou dit er dan uit zien?

EECOLOR replied:
Als we hier (server-side) iets zouden willen doen met de gebruiker, dan moeten we ook het `req` gebruiken. In de meeste gevallen staat er iets in de `header` of `cookie` waarmee we dan de gebruiker kunnen identificeren.

larsvankleef commented on example/src/advanced/routeMap.js:
```diff
@@ -0,0 +1,59 @@
+      path: { nl: 'artikelen', en: 'articles' },
+      data: { title: { nl: 'Artikelen', en: 'Articles' } },
+
+      list: {
```
Eigenlijk is `articles` al de `list` view? Zou het niet logische zijn als we dit niet in `list` zetten maar in de data van `articles`

EECOLOR replied:
Goede vraag en dit is een hele bewuste keuze.

Wanneer we een artikel tonen is het niet nodig om alle artikelen op te halen, dat zou wel gebeuren wanneer `list.data` direct in `articles` zou staan. Door hier een losse route van te maken hebben we de vrijheid om te kiezen.

Het patroon wat je voorstelt hebben we wel gebruikt bij `article`: voor elke `tab` moet het artikel aanwezig zijn.

In de FAQ hebben we dit staan:

> - Why is the route map itself not a route?
>   - It would make it impossible to have home as a route that is not the parent of any other routes. This makes some data fetching patterns impossible.

larsvankleef commented on example/src/advanced/routeMap.js:
```diff
@@ -0,0 +1,59 @@
+        path: ':articleId',
+        data: async ({ articleId }) => ({ article: await fetchArticle({ articleId }) }),
+
+        main: '',
```
Weet niet zo goed wat ik vind van de lege strings. Zou zelf misschien meer gaan voor '/'

EECOLOR replied:
De keuze voor lege strings is een bewuste keuze. Een aantal redenen:

- Het is makkelijker uit te leggen. Wanneer je een `'/'` gebruikt kom je in de knoop met andere routes, is het dan `'/:articleId'`, `':articleId/'` of `'/:articleId/'`? Ik wil liever niet meerdere varianten ondersteunen.
- Het geeft minder 'noise'
- Het zorgt ervoor dat de implementatie simpeler is

larsvankleef commented on package.json:
```diff
@@ -0,0 +1,34 @@
+    "Erik Westra <erik@kaliber.net>"
+  ],
+  "scripts": {
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest"
```
Yeah JEST!

Ik heb laatst deze cursus afgerond: https://testingjavascript.com/

EECOLOR replied:
Als je het leuk vindt: uiteindelijk wil ik de React functies in `routing.js` ook nog testen. Ik had alleen niet de puf om dat nog toe te voegen.

larsvankleef commented on readme.md:
```diff
@@ -0,0 +1,391 @@
+    [routeMap.notFound, { status: 404 }],
+  )
+  return 'route' in result
+    ? { status: result.status, data: await route.data(params) }
```
Dit staat een beetje haaks op wat je zult vinden in de advanced example. Daar is een verschil tussen de data op de route en de async api data gemaakt.

EECOLOR replied:
Tja, blijft tricky. Dit is een totaal onvolledig voorbeeld om een beetje een idee te geven wat er mogelijk is. Data fetching is geen onderdeel van de library en is enorm afhankelijk van het front-end framework dat gebruikt wordt.

De variant van `advanced` is er 1 die zowel server als client side ondersteunt, die complixiteit wil je niet als eerste in de readme zien.

larsvankleef reviewed (commented):
Cool! Ik heb vooral vragen en ik vond de API die gemaakt is wat lastig om in een keer te snappen, maar met de voorbeelden werd het al wat duidelijker. 

Leuk om nog even over te bellen! Dan kan ik mijn opmerkingen ook wat meer toe lichten.

EECOLOR commented:
@larsvankleef Onwijs bedankt voor je review. We moeten zeker nog even een keer een moment prikken om je opmerkingen te bespreken.

EECOLOR committed:
Simplify fetching data on match

EECOLOR committed:
Improve `usePick` documentation

EECOLOR committed:
Rename `route` to `matchRoute` and `routes` to `matchRoutes`

- Changed some internals to simplify internal structure

EECOLOR committed:
Rename `seconds` to `resolveAfter`

EECOLOR committed:
Simplify route data fetching in example

EECOLOR committed:
Rename useRoute to useMatchedRoute

EECOLOR committed:
Added convenience function to obtain route data

EECOLOR commented:
https://github.com/kaliberjs/build/pull/219

EECOLOR committed:
Small refactoring of exampes

EECOLOR commented:
One of the open points is: 'more realistic examples'. This is quite a hard thing to do. So what I will do it merge this first version and release it as a `0.0.1` version. That allows us to use it in some other projects to figure out what a real example would like (when used with something like Sanity).
